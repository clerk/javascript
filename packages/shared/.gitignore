browser
callWithRetry
color
cookie
date
deprecated
error
file
globs
handleValueOrFn
isomorphicAtob
isomorphicBtoa
keys
loadScript
localStorageBroadcastChannel
poller
proxy
underscore
url
react
constants
apiUrlFromPublishableKey