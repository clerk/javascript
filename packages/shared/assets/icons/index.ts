// @ts-nocheck
/**
 * TypeScript configuration (typings) is not correctly configured for all projects.
 * Consequently, the files are correctly imported but the TS checker emits errors.
 * The above no-check is safe, as webpack will not allow compilation if for example a file is not resolved.
 */
export { default as ArrowLeftIcon } from './arrow-left.svg';
export { default as ArrowRightIcon } from './arrow-right.svg';
export { default as BinIcon } from './bin.svg';
export { default as CameraIcon } from './camera.svg';
export { default as CaretIcon } from './caret.svg';
export { default as CheckCircleIcon } from './check-circle.svg';
export { default as CheckmarkIcon } from './checkmark.svg';
export { default as CogIcon } from './cog.svg';
export { default as CornerDownRightIcon } from './corner-down-right.svg';
export { default as DesktopIcon } from './desktop.svg';
export { default as DropletIcon } from './droplet.svg';
export { default as ExclamationTriangle } from './exclamation-triangle.svg';
export { default as ImageDefaultIcon } from './image-default.svg';
export { default as LoaderIcon } from './loader.svg';
export { default as LogoMarkIcon } from './logo-mark.svg';
export { default as MailIcon } from './mail.svg';
export { default as MailOpen } from './mail-open.svg';
export { default as MobileIcon } from './mobile.svg';
export { default as MoreVerticalIcon } from './more-vertical.svg';
export { default as PhoneIcon } from './phone.svg';
export { default as ServerIcon } from './server.svg';
export { default as ShieldIcon } from './shield.svg';
export { default as SignOutIcon } from './sign-out.svg';
export { default as SwitchArrows } from './switch-arrows.svg';
export { default as TickShield } from './tick-shield.svg';
export { default as UploadIcon } from './upload.svg';
export { default as UserIcon } from './user.svg';
export { default as UserPlusIcon } from './user-plus.svg';
