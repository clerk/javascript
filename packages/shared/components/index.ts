export * from './button';
export * from './card';
export * from './checkbox';
export * from './dropdown';
export * from './form';
export * from './input';
export * from './label';
export * from './modal';
export * from './outline';
export * from './popover';
export * from './portal';
export * from './radio';
export * from './spinner';
export * from './switch';
export * from './tabs';
export * from './titledCard';
export * from './toggle';
export * from './tooltip';
export * from './wizard';
export * from './heading';
export * from './oneTimeCodeInput';
export * from './tag';
export * from './textArea';
export * from './list';
export * from './avatar';
export * from './phoneInput';
export * from './control';
export * from './imageUploader';
