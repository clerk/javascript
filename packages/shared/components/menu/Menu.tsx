import cn from 'classnames';
import React from 'react';

// @ts-ignore
import styles from './Menu.module.scss';

export type MenuOption = {
  label: React.ReactNode;
  icon?: React.ReactNode;
  isDestructiveAction?: boolean;
  handleSelect?: (o: Record<string, unknown>) => void;
};

export type MenuProps = {
  options: MenuOption[];
  children?: React.ReactNode;
  className?: string;
  style?: any;
};

export const Menu: React.FC<MenuProps> = ({ className, options, style }: MenuProps) => {
  return (
    <ul
      role='menu'
      className={cn(styles.menu, className)}
      style={style}
    >
      {options.map((option, i) => {
        const onClick = (e: React.MouseEvent<HTMLLIElement>) => {
          e.stopPropagation();
          if (typeof option.handleSelect === 'function') {
            option.handleSelect(option);
          }
        };

        return (
          <li
            key={i}
            role='menuitem'
            className={cn(styles.menuItem, {
              [styles.destructiveAction]: option.isDestructiveAction,
            })}
            onClick={onClick}
            tabIndex={0}
          >
            {option.icon && <div className={styles.iconContainer}>{option.icon}</div>}
            <div>{option.label}</div>
          </li>
        );
      })}
    </ul>
  );
};
