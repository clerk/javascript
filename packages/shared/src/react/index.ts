export * from './hooks';

export {
  ClerkInstanceContext,
  ClientContext,
  OptionsContext,
  OrganizationProvider,
  SessionContext,
  useAssertWrappedByClerkProvider,
  useClerkInstanceContext,
  useClientContext,
  useOptionsContext,
  useOrganizationContext,
  UserContext,
  useSessionContext,
  useUserContext,
} from './contexts';
