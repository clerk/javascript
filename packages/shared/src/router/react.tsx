/**
 * React-specific binding's for interacting with Clerk's router interface.
 */
import React, { createContext, useContext } from 'react';

import type { ClerkHostRouter, ClerkRouter } from './router';
import { createClerkRouter } from './router';

export const ClerkHostRouterContext = createContext<ClerkHostRouter | null>(null);
export const ClerkRouterContext = createContext<ClerkRouter | null>(null);

export function useClerkHostRouter() {
  const ctx = useContext(ClerkHostRouterContext);

  if (!ctx) {
    throw new Error(
      'clerk: Unable to locate ClerkHostRouter, make sure this is rendered within `<ClerkHostRouterContext.Provider>`.',
    );
  }

  return ctx;
}

export function useClerkRouter() {
  const ctx = useContext(ClerkRouterContext);

  if (!ctx) {
    throw new Error('clerk: Unable to locate ClerkRouter, make sure this is rendered within `<Router>`.');
  }

  return ctx;
}

/**
 * Construct a Clerk Router using the provided host router. The router instance is accessible using `useClerkRouter()`.
 */
export function Router({
  basePath,
  children,
  router,
}: {
  children: React.ReactNode;
  basePath?: string;
  router?: ClerkHostRouter;
}) {
  const hostRouter = useClerkHostRouter();
  const clerkRouter = createClerkRouter(router ?? hostRouter, basePath);

  return <ClerkRouterContext.Provider value={clerkRouter}>{children}</ClerkRouterContext.Provider>;
}

type RouteProps = { path?: string; index?: boolean };

/**
 * Used to conditionally render its children based on whether or not the current path matches the provided path.
 */
export function Route({ path, children, index }: RouteProps & { children: React.ReactNode }) {
  const parentRouter = useClerkRouter();

  if (!path && !index) {
    return children;
  }

  if (!parentRouter?.match(path, index)) {
    return null;
  }

  return children;
}
