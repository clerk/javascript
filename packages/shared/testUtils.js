module.exports = require('./utils/testUtils');
