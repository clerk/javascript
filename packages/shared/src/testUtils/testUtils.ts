import '@testing-library/jest-dom/extend-expect';

import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';

import { noop } from '../utils';

// Wrap components with test providers
const customRender = (ui: React.ReactElement, options?: RenderOptions) => {
  return render(ui, options);
};

// Re-export everything
export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

// Override render method
export { customRender as render, noop };
