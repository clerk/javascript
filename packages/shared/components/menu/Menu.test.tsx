import { render, renderJSON, screen, userEvent } from '@clerk/shared/testUtils';
import * as React from 'react';

import { ServerIcon } from '../../assets/icons';
import { Menu } from './Menu';

describe('<Menu/>', () => {
  const options = [
    {
      label: 'Foo',
      icon: <ServerIcon />,
    },
    { label: 'Bar' },
  ];

  it('renders the menu options', () => {
    const tree = renderJSON(
      <Menu
        options={options}
        className='foo'
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('triggers the option callback upon clicking on it', async () => {
    const opts = options.map(o => ({
      ...o,
      handleSelect: jest.fn(),
    }));

    render(<Menu options={opts} />);

    const trigger1 = screen.getByText('Foo');
    await userEvent.click(trigger1);
    expect(opts[0].handleSelect).toHaveBeenCalledTimes(1);

    const trigger2 = screen.getByText('Bar');
    await userEvent.click(trigger2);
    expect(opts[0].handleSelect).toHaveBeenCalledTimes(1);
  });
});
