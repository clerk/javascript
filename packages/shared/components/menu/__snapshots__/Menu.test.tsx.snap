// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu/> renders the menu options 1`] = `
<ul
  className="menu foo"
  role="menu"
>
  <li
    className="menuItem"
    onClick={[Function]}
    role="menuitem"
    tabIndex={0}
  >
    <div
      className="iconContainer"
    >
      <svg
        data-filename="assets/icons/server.svg"
      />
    </div>
    <div>
      Foo
    </div>
  </li>
  <li
    className="menuItem"
    onClick={[Function]}
    role="menuitem"
    tabIndex={0}
  >
    <div>
      Bar
    </div>
  </li>
</ul>
`;
